bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -1446,6 +1304,153 @@ public LedgerStorage getLedgerStorage() {
+                return;
+            }
+
+            if (!bookieStatus.setToWritableMode()) {
+                // do nothing if already in writable mode
+                return;
+            }
+            LOG.info("Transitioning Bookie to Writable mode and will serve read/write requests.");
+            if (conf.isPersistBookieStatusEnabled()) {
+                bookieStatus.writeToDirectories(ledgerDirsManager.getAllLedgerDirs());
+            }
+            // change zookeeper state only when using zookeeper
+            if (null == registrationManager) {
+                return;
+            }
+            try {
+                doRegisterBookie(false);
+            } catch (IOException e) {
+                LOG.warn("Error in transitioning back to writable mode : ", e);
+                transitionToReadOnlyMode();
+                return;
+            }
+            // clear the readonly state
+            try {
+                registrationManager.unregisterBookie(bookieId, true);
+            } catch (BookieException e) {
+                // if we failed when deleting the readonly flag in zookeeper, it is OK since client would
+                // already see the bookie in writable list. so just log the exception
+                LOG.warn("Failed to delete bookie readonly state in zookeeper : ", e);
+                return;
+            }
+        }
+        @VisibleForTesting
+        public void doTransitionToReadOnlyMode() {
+            if (shuttingdown) {
+                return;
+            }
+            if (!bookieStatus.setToReadOnlyMode()) {
+                return;
+            }
+            if (!conf.isReadOnlyModeEnabled()) {
+                LOG.warn("ReadOnly mode is not enabled. "
+                        + "Can be enabled by configuring "
+                        + "'readOnlyModeEnabled=true' in configuration."
+                        + "Shutting down bookie");
+                triggerBookieShutdown(ExitCode.BOOKIE_EXCEPTION);
+                return;
+            }
+            LOG.info("Transitioning Bookie to ReadOnly mode,"
+                    + " and will serve only read requests from clients!");
+            // persist the bookie status if we enable this
+            if (conf.isPersistBookieStatusEnabled()) {
+                this.bookieStatus.writeToDirectories(ledgerDirsManager.getAllLedgerDirs());
+            }
+            // change zookeeper state only when using zookeeper
+            if (null == registrationManager) {
+                return;
+            }
+            try {
+                registrationManager.registerBookie(bookieId, true);
+            } catch (BookieException e) {
+                LOG.error("Error in transition to ReadOnly Mode."
+                        + " Shutting down", e);
+                triggerBookieShutdown(ExitCode.BOOKIE_EXCEPTION);
+                return;
+            }
+        }
+
+    }
+
+    @VisibleForTesting
+    public StateManager getStateManager() {
```

Comment:
do you need to make it public?

Reply:
you can return BookieStateManager here? so you don't need to cast in the test cases.

Reply:
because tests in `org.apache.bookkeeper.test` may use stateManager, for example ReadOnlyBookieTest, so `getStateManager ` should use public